File: Homework/Objective-C/HW9/HW9/Patient.m
//
//  Patient.m
//  HW9
//
//  Created by Руслан on 21.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "Patient.h"

@implementation Patient

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.temperature = (arc4random() % 60) / 10 + 36.6f;
    }
    return self;
}

- (BOOL) howAreYou {
    
    BOOL isOk = arc4random() % 2;
    
    if (!isOk) {
        self.symptom = arc4random() % 3;
        [self.delegate patientFeelsBad:self];
        self.feedback = arc4random() % 2;
    } else {
        NSLog(@"%@ is completely healthy", self.name);
    }
    
    return isOk;
    
}

- (void) change:(id<PatientDelegate>)doctor from:(NSArray *)doctors {
    
    for (id<PatientDelegate> newDoctor in doctors) {
        if (newDoctor != doctor) {
            self.delegate = newDoctor;
            NSLog(@"Patient's %@ has a new doctor. He(She) is %@", self.name, newDoctor.name);
            break;
        }
    }
}

- (void) takePill {
    NSLog(@"%@ takes a pill :)", self.name);
}

- (void) makeShot {
    NSLog(@"%@ make a shot", self.name);
}

@end

File: Homework/Swift/HW9/HW9/Patient.swift
//
//  Patient.swift
//  HW9
//
//  Created by Руслан on 22.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class Patient: NSObject {

    enum Symptoms {
        case soreThroat
        case stomachache
        case headache
    }
    
    enum Feedback: Int {
        case like
        case dislike
    }
    
// MARK: - Patient properties
    
    let name: String
    var temperature: Float
    var delegate: PatientDelegate!
    var symptom: Symptoms!
    var feedback: Feedback!
    
// MARK: - Patient init
    
    init(name: String) {
        self.name = name
        self.temperature = Float(Int(arc4random()) % 50) / 10 + 36
    }
    
// MARK: - Patient methods
    
    func howAreYou() {
        
        let isOk: Bool = arc4random() % 2 == 1 ? true : false;
        
        if !isOk {
            let intSymptom = arc4random() % 3;
            
            switch intSymptom {
            case 0:
                self.symptom = .soreThroat
            case 1:
                self.symptom = .stomachache
            case 2:
                self.symptom = .headache
            default: break
            }
            
            self.delegate.patientFeelsBad(patient: self)
            
            self.feedback = (arc4random() % 2) == 0 ? .like : .dislike
        } else {
            print("\(self.name) is completely healthy", terminator: "\n\n")
        }
    }
    
    func takePill() {
        print("\(self.name) takes a pill")
    }
    
    func makeShot() {
        print("\(self.name) make a shot")
    }
    
    func change(doctor: PatientDelegate, from doctors: [PatientDelegate]) {
        
        for newDoctor in doctors {
            if doctor.name != newDoctor.name {
                self.delegate = newDoctor
                print("Patient \(self.name) has a new doctor. His(Her) name is \(newDoctor.name)")
                break
            }
        }
    }
}

// MARK: - PatientDelegate

protocol PatientDelegate {
    
    var name: String { get }
    
    func patientFeelsBad(patient: Patient)
    
}
